### PR TITLE
[SPARK-34446][SS][DOCS] Update doc for stream-stream join (full outer + left semi)

### PR DESCRIPTION
**Pull request by c21**
### What changes were proposed in this pull request?

Per discussion in https://issues.apache.org/jira/browse/SPARK-32883?focusedCommentId=17285057&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17285057, we should add documentation for added new features of full outer and left semi joins into SS programming guide.

* Reworded the section for "Outer Joins with Watermarking", to make it work for full outer join. Updated the code snippet to show up full outer and left semi join.
* Added one section for "Semi Joins with Watermarking", similar to "Outer Joins with Watermarking".
* Updated "Support matrix for joins in streaming queries" to reflect latest fact for full outer and left semi join.

### Why are the changes needed?

Good for users and developers to follow guide to try out these two new features.

### Does this PR introduce _any_ user-facing change?

Yes. They will see the corresponding updated guide.

### How was this patch tested?

No, just documentation change. Previewed the markdown file in browser.
Also attached here for the change to the "Support matrix for joins in streaming queries" table.

Screen Shot 2021-02-16 at 8 12 07 PM

### CODE REVIEW

**Comment by c21** [comment]
cc @HeartSaVioR , @cloud-fan , @HyukjinKwon , @xuanyuanking and @gatorsmile , thanks.

**Comment by HeartSaVioR** [comment]
Thanks for the quick turnaround. I'll take a look tomorrow - hard to concentrate today.

cc. @viirya as he reviewed the series of PRs as well.

File: docs/structured-streaming-programming-guide.md [hunk 1]
```diff
@@ -1337,7 +1337,7 @@ impressionsWithWatermark.join(
     clickTime >= impressionTime AND
     clickTime <= impressionTime + interval 1 hour
     """),
-  joinType = "leftOuter"      // can be "inner", "leftOuter", "rightOuter"
+  joinType = "leftOuter"      // can be "inner", "leftOuter", "rightOuter", "fullOuter", "leftSemi"
```
Comment (viirya) [inline docs/structured-streaming-programming-guide.md:1340]: Not due to this PR, but I saw this section is "Outer Joins with Watermarking", seems to be "inner" and "leftSemi" should not be here.

Comment (HeartSaVioR) [reply docs/structured-streaming-programming-guide.md:1340]: probably the purpose was to enumerate the all possible values for joinType. while I agree it's not matching with the section, I feel it's not a big deal. either is fine, including leaving it as it is (with adding left semi) and another minor PR to fix.

File: docs/structured-streaming-programming-guide.md [hunk 2]
```diff
@@ -1415,6 +1415,18 @@ generation of the outer result may get delayed if there no new data being receiv
 *In short, if any of the two input streams being joined does not receive data for a while, the
 outer (both cases, left or right) output may get delayed.*
 
+##### Semi Joins with Watermarking
+A semi join returns values from the left side of the relation that has a match with the right.
+It is also referred to as a left semi join. Similar to outer joins, watermark + event-time
+constraints must be specified for semi join. This is because for not generating result for an input
```
Comment (HeartSaVioR) [inline docs/structured-streaming-programming-guide.md:1421]: I see the sentence is a slightly modified one from outer join, but doesn't seem to be intuitive for left semi case. Watermark constraint wouldn't be a requirement for left semi join if we allow Spark to buffer entire input rows in state, and the result would be the same. (Please correct me if I'm missing here.)

The main purpose is to evict input rows on left side which are no longer possibly matching with the right side, so we'd be better emphasizing the eviction as that's the main purpose. What about

"This is to evict unmatched input rows on left side, the engine must know when an input row on left side is not going to match with anything on right side in future."

**Review by HeartSaVioR — COMMENTED** [review]
Looks OK, one comment.

Probably helpful for reviewers if you could capture the screenshot of the change, especially the table side. I won't ask about the text change but it includes the changes on table element which is easy to break and also easy to miss on review phase.

**Comment by c21** [comment]
@HeartSaVioR - updated the description for required table screenshot and addressed comment of rewording in semi join section.

File: docs/structured-streaming-programming-guide.md (hunk 1 above)
Comment (c21) [reply docs/structured-streaming-programming-guide.md:1340]: I don't feel this is too big deal, leave it as it is for now.

**Review by HeartSaVioR — APPROVED** [review]
+1

**Comment by HeartSaVioR** [comment]
Thanks for the quick fix, @c21 !
I'll merge this to master/branch-3.1 in tomorrow morning in KST unless there's no further comment.

**Comment by HeartSaVioR** [comment]
OK no further comments. Merging to master/branch-3.1!

**Comment by HyukjinKwon** [comment]
Thanks @HeartSaVioR!

**Comment by c21** [comment]
Thanks @HeartSaVioR and @viirya for quick review!

**Pull request closed without merging**